### PR TITLE
Consolidate remaining parquet config options into ConfigOptions

### PR DESCRIPTION
**Pull request by alamb**
~Draft as it builds on https://github.com/apache/arrow-datafusion/pull/3822~

# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/3821

 # Rationale for this change
1. Make it easier for people to see what config options are available will make it more likely they are used
2. The more mechanisms that configuration is supplied, the more likely it to confuse people

It turns out options for reading parquet files were able to be set (and possibly) overridden by no less than three different structures! This is confusing, to say the least. 

This is especially important to us as we move to use more and more of the parquet code directly in IOx (e.g. https://github.com/influxdata/influxdb_iox/issues/5887 and https://github.com/influxdata/influxdb_iox/issues/5897)

This also helps towards #3887 

# What changes are included in this PR?
1. move metadata_size_hint, enable_pruning, and merge_schema_metadata to new config options
2. Remove redundancy and config overrides

# Are there any user-facing changes?
The main change is that now the parquet reader settings are session wide. 

Previously, depending on which of the APIs was used to create / register / run parquet, the settings might change if you change the session config or they might have been a snapshot based on when you registered the reader

### CODE REVIEW

**Comment by alamb** [comment]
There are still some tests and code updates required for this PR

File: datafusion/core/src/datasource/file_format/parquet.rs [hunk 1]
```diff
@@ -52,59 +57,70 @@ pub const DEFAULT_PARQUET_EXTENSION: &str = ".parquet";
 /// The Apache Parquet `FileFormat` implementation
 #[derive(Debug)]
 pub struct ParquetFormat {
-    enable_pruning: bool,
```
Comment (alamb) [inline datafusion/core/src/datasource/file_format/parquet.rs:55]: Here was one copy of (some) of the settings

File: datafusion/core/src/execution/context.rs [hunk 2]
```diff
@@ -1142,8 +1146,6 @@ pub struct SessionConfig {
     /// Should DataFusion repartition data using the partition keys to execute window functions in
     /// parallel using the provided `target_partitions` level
     pub repartition_windows: bool,
-    /// Should DataFusion parquet reader using the predicate to prune data
-    pub parquet_pruning: bool,
```
Comment (alamb) [inline datafusion/core/src/execution/context.rs:1146]: here is a second copy of one of the settings

File: datafusion/core/src/execution/options.rs [hunk 3]
```diff
@@ -168,56 +170,31 @@ pub struct ParquetReadOptions<'a> {
     pub file_extension: &'a str,
     /// Partition Columns
     pub table_partition_cols: Vec<String>,
-    /// Should DataFusion parquet reader use the predicate to prune data,
```
Comment (alamb) [inline datafusion/core/src/execution/options.rs:171]: here is a third copy, again of some of the settings

File: datafusion/core/tests/sql/information_schema.rs [hunk 4]
```diff
@@ -702,8 +702,11 @@ async fn show_all() {
         "| datafusion.execution.coalesce_batches           | true    |",
         "| datafusion.execution.coalesce_target_batch_size | 4096    |",
         "| datafusion.execution.parquet.enable_page_index  | false   |",
+        "| datafusion.execution.parquet.metadata_size_hint | NULL    |",
```
Comment (alamb) [inline datafusion/core/tests/sql/information_schema.rs:705]: defaults are now even clearer!

File: datafusion/core/tests/sql/parquet_schema.rs [hunk 5]
```diff
@@ -141,6 +139,13 @@ async fn schema_merge_can_preserve_metadata() {
     let table_path = table_dir.to_str().unwrap().to_string();
 
     let ctx = SessionContext::new();
+
+    // explicitly disable schema clearing
+    ctx.config_options()
```
Comment (alamb) [inline datafusion/core/tests/sql/parquet_schema.rs:144]: Ideally we will have a nicer API for this: https://github.com/apache/arrow-datafusion/issues/3908

File: datafusion/core/src/config.rs [hunk 6]
```diff
@@ -237,6 +247,29 @@ impl BuiltInConfigs {
...
+                 multiple parquet files with schemas containing compatible types but different metadata.",
+                true,
+            ),
+            ConfigDefinition::new(
+                OPT_PARQUET_METADATA_SIZE_HINT,
+                "If specified, the parquet reader will try and fetch the last `size_hint` \
+                 bytes of the parquet file optimistically. If not specified, two read are required: \
+                 One read to fetch the 8-byte parquet footer and  \
+                 another to fetch the metadata length encoded in the footer.",
+                DataType::Boolean,
```
Comment (tustvold) [inline datafusion/core/src/config.rs:270]: Why is this a boolean?

**Review by tustvold — COMMENTED** [review]
Honestly I'm really struggling to review this PR as I'm honestly lost trying to understand the config system...

We used to have just `RuntimeConfig`/ `RuntimeEnv` for process global settings, `SessionConfig`/`SessionContext` for session local settings i.e. planning, and `TaskContext` for physical plan execution settings

I'm not entirely clear on how `ConfigOptions` fits into this, what scope it has, etc... It appears to have been introduced by @andygrove in #2754 with the intention articulated on #2756 to use it for `SessionContext`

Plumbing this down into the physical plans therefore feels a bit like it is coupling together parts of the system that shouldn't be coupled together? I would have expected the physical operators to only be exposed to `TaskContext`, with `SessionContext` solely used for planning.

File: datafusion/core/src/execution/options.rs [hunk 7]
```diff
@@ -168,56 +170,31 @@ pub struct ParquetReadOptions<'a> {
...
             table_partition_cols: vec![],
-            parquet_pruning: format_default.enable_pruning(),
-            skip_metadata: format_default.skip_metadata(),
         }
     }
 }
 
 impl<'a> ParquetReadOptions<'a> {
-    /// Specify parquet_pruning
-    pub fn parquet_pruning(mut self, parquet_pruning: bool) -> Self {
```
Comment (tustvold) [inline datafusion/core/src/execution/options.rs:197]: Doesn't this represent a functionality loss, previously you could configure these options on per-datasource basis. Now they are located on ConfigOptions which is session global?

File: datafusion/core/src/config.rs (hunk 6 above)
Comment (alamb) [reply datafusion/core/src/config.rs]: You are right -- it should be u64 -- fixed in e5e3be076

**Comment by alamb** [comment]
> I'm really struggling to review this PR as I'm honestly lost trying to understand the design of the config system...

Yes, I agree it is very confusing. 

Would you find it less confusing if the physical planner got a `SessionConfig` rather than `ConfigOptions`?


> Plumbing this down into the physical plans therefore feels a bit like it is coupling together parts of the system that shouldn't be coupled together? I would have expected the physical operators to only be exposed to TaskContext, with SessionContext solely used for planning.

I think the `SessionContext` has a `SessionState` which has  a `SessionConfig` which has a `ConfigOptions` 🤯 

TaskContext has `TaskProperties` which is either a `SessionConfig` or some k=v pairs

So in that view of things  your expectation "expected the physical operators to only be exposed to TaskContext, with SessionContext solely used for planning." does hold

In my mind `ConfigOptions` and `SessionConfig` serve the same purpose and eventually they will be unified (using the implementation of `ConfigOptions`.

**Comment by alamb** [comment]
Here is my plan: I will write up a ticket explaining the current state of affairs in terms of config (and I will make a diagram) and we can use that to come to consensus on where we want to head. 

Given the current mess of parquet options (has 3 *other* places to place config values *in addition* to the SessonConfig / ConfigOptions I think this PR is an improvement anyways.

We'll see if @andygrove  has any thoughts

File: datafusion/core/src/execution/options.rs (hunk 7 above)
Comment (alamb) [reply datafusion/core/src/execution/options.rs:197]: I would say "kind of" -- previously you could configure the options on a per-datasource basis but depending on exactly what codepath you used and what other options were set your settings might or might not get overridden.

**Comment by alamb** [comment]
Marking as a draft until I get a writeup and some consensus

**Comment by alamb** [comment]
Here is a writeup of a way to improve the configuration situation somewhat https://github.com/apache/arrow-datafusion/issues/4349

File: datafusion/core/src/config.rs [hunk 8]
```diff
@@ -237,6 +247,29 @@ impl BuiltInConfigs {
...
+                true,
+            ),
+            ConfigDefinition::new(
+                OPT_PARQUET_METADATA_SIZE_HINT,
+                "If specified, the parquet reader will try and fetch the last `size_hint` \
+                 bytes of the parquet file optimistically. If not specified, two read are required: \
+                 One read to fetch the 8-byte parquet footer and  \
+                 another to fetch the metadata length encoded in the footer.",
+                DataType::UInt64,
+                ScalarValue::UInt64(None),
```
[the remark this answers is not in the document]
Comment (alamb) [reply datafusion/core/src/config.rs:271]: 64K seems reasonable to me

**Comment by alamb** [comment]
FWIW My plan for this PR is:
1. To leave any the current places to configure parquet settings per ParquetExec as overrrides (`Option<val>`) 
2. Expose remaining options via `ConfigOptions` (and make them visible, etc)

File: datafusion/core/src/config.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (alamb) [reply datafusion/core/src/config.rs:271]: Upon more thought I would like to propose we do the "consolidate config code" and "change default value" in separate PRs (which I think will be easier to see / review once the configuration is consolidated)

**Comment by alamb** [comment]
Updated version in https://github.com/apache/arrow-datafusion/pull/4427

**Pull request closed without merging**